File: python/tvm/op.py
from __future__ import absolute_import as _abs

_binary_op_cls = None

class BinaryOp(object):
    """Base class of binary operator"""
    def __call__(self, lhs, rhs):
        return _binary_op_cls(self, lhs, rhs)

class AddOp(BinaryOp):
    pass

class SubOp(BinaryOp):
    pass

class MulOp(BinaryOp):
    pass

class DivOp(BinaryOp):
    pass


add = AddOp()
sub = SubOp()
mul = MulOp()
div = DivOp()

File: src/op.cc
/*!
 *  Copyright (c) 2016 by Contributors
 * \file op.cc
 */
#include <tvm/op.h>
#include <tvm/expr_node.h>

namespace tvm {

Expr BinaryOp::operator()(Expr lhs, Expr rhs) const {
  auto nptr = std::make_shared<BinaryOpNode>(
      this, std::move(lhs), std::move(rhs));
  nptr->Verify();
  return Expr(std::move(nptr));
}

#define DEFINE_SINGLETON_GET(TypeName)           \
  TypeName* TypeName::Get() {                    \
    static TypeName inst;                        \
    return &inst;                                \
  }

DEFINE_SINGLETON_GET(AddOp);
DEFINE_SINGLETON_GET(SubOp);
DEFINE_SINGLETON_GET(MulOp);
DEFINE_SINGLETON_GET(DivOp);
DEFINE_SINGLETON_GET(MaxOp);
DEFINE_SINGLETON_GET(MinOp);

}  // namespace tvm
